### PR TITLE
feat: ability for container and custom types to have nested elements

### PR DESCRIPTION
**Pull request by lizaiv77**
Grammar changes reflected in doc https://github.com/salesforce-experience-platform-emu/utam-docs/pull/985
Implementation required to create getter that returns basic type instead of container and custom, for container we also needed to add field with proper locator.
For example after this change if container `rootContainer` element has nested elements, we now also create field and private method that returns container element as basic scope:
```java
@ElementMarker.Find(css = "slot")
 private ElementLocation rootContainerScope;

final BasicElement getRootContainerScopeElement() {
    BasicElement root = this.getRootElement();
    return basic(root, this.rootContainerScope).build(BasicElement.class, BasePageElement.class);
  }
```

Similarly if custom element `customRoot` has nested elements, we create private getter to return custom element as basic scope (field for custom element is created anyway): 
```java
final BasicElement getCustomRootScopeElement() {
    BasicElement root = this.getRootElement();
    return basic(root, this.customRoot).build(BasicElement.class, BasePageElement.class);
 }
```

### CODE REVIEW

File: utam-compiler/src/test/resources/validate/custom_element/nestedElementsForList.json [hunk 1]
```diff
@@ -4,13 +4,14 @@
...
+        "css": "css",
+        "returnAll": true
       },
-      "elements" : [
+      "elements": [
         {
-          "name": "nested",
-          "selector": {
-            "css": "css2"
+          "name" : "nested",
```
Comment (jimevanssfdc) [inline utam-compiler/src/test/resources/validate/custom_element/nestedElementsForList.json:12]: What's our consistency here for JSON formatting? Do we expect a space before the colon or not? I don't particularly care, but let's at least make an attempt to be consistent within a single file. :D

File: utam-compiler/src/test/java/utam/compiler/grammar/UtamElement_FrameTests.java [hunk 2]
```diff
@@ -71,25 +71,6 @@ public void testFrameElementWithNullableThrows() {
                 + "supported are: name, public, selector, type"));
   }
 
-  @Test
-  public void testFrameElementWithElementsThrows() {
```
Comment (jimevanssfdc) [inline utam-compiler/src/test/java/utam/compiler/grammar/UtamElement_FrameTests.java:75]: Wouldn't this still be a valid test? According to the docs PR, frames still can't have nested elements. Or did I miss where this is tested elsewhere (a distinct possibility)?

**Review by jimevanssfdc — APPROVED** [review]

File: utam-compiler/src/test/resources/validate/custom_element/nestedElementsForList.json (hunk 1 above)
Comment (lizaiv77) [reply utam-compiler/src/test/resources/validate/custom_element/nestedElementsForList.json:12]: we have spotless to take care of that

File: utam-compiler/src/test/java/utam/compiler/grammar/UtamElement_FrameTests.java (hunk 2 above)
Comment (lizaiv77) [reply utam-compiler/src/test/java/utam/compiler/grammar/UtamElement_FrameTests.java:75]: refactored in renamed class

File: utam-compiler/src/main/java/utam/compiler/helpers/TranslationContext.java [hunk 3]
```diff
@@ -251,7 +251,27 @@ public List<PageObjectMethod> getMethods() {
    * @return the list of fields defined in the Page Object
    */
   public List<PageClassField> getFields() {
-    return pageObjectFields;
+    Set<String> uniqueFieldNames = new HashSet<>();
```
[the remark this answers is not in the document]
Comment (lizaiv77) [reply utam-compiler/src/main/java/utam/compiler/helpers/TranslationContext.java:254]: we could, but it does not change code below much

**Pull request merged**